hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java

```diff
@@ -707,20 +709,33 @@ protected List<String> getInstantsToRollback(HoodieTableMetaClient metaClient, H
         }
       }).collect(Collectors.toList());
     } else if (cleaningPolicy.isLazy()) {
-      return inflightInstantsStream.filter(instant -> {
-        try {
-          return heartbeatClient.isHeartbeatExpired(instant.getTimestamp());
-        } catch (IOException io) {
-          throw new HoodieException("Failed to check heartbeat for instant " + instant, io);
-        }
-      }).map(HoodieInstant::getTimestamp).collect(Collectors.toList());
+      return getInstantsToRollbackForLazyCleanPolicy(metaClient, inflightInstantsStream);
     } else if (cleaningPolicy.isNever()) {
       return Collections.emptyList();
     } else {
       throw new IllegalArgumentException("Invalid Failed Writes Cleaning Policy " + config.getFailedWritesCleanPolicy());
     }
   }
 
+  @VisibleForTesting
+  public List<String> getInstantsToRollbackForLazyCleanPolicy(HoodieTableMetaClient metaClient,
+                                                              Stream<HoodieInstant> inflightInstantsStream) {
+    // Get expired instants, must store them into list before double-checking
+    List<String> expiredInstants = inflightInstantsStream.filter(instant -> {
+      try {
+        // An instant transformed from inflight to completed have no heartbeat file and will be detected as expired instant here
+        return heartbeatClient.isHeartbeatExpired(instant.getTimestamp());
+      } catch (IOException io) {
+        throw new HoodieException("Failed to check heartbeat for instant " + instant, io);
+      }
+    }).map(HoodieInstant::getTimestamp).collect(Collectors.toList());
+
+    // Double check whether the heartbeat-expired instant is an inflight instant
+    metaClient.reloadActiveTimeline();
```

Comment:
even reload active timeline, there is still possibility to rollback completed instant if the instant completed after ` metaClient.reloadActiveTimeline`?

Reply:
Only heartbeat-expired instants will be rollback. If the instant is finished after `metaClient.reloadActiveTimeline`,  the heartbeat is not expired when checking heartbeat, so it won't be rollback. That why we must check heartbeat first and then reload the timeline to do the double check

Reply:
active timeline reload is an expensive operation. we should try to minimize. 
can you check if exporedInstants is not empty and reload only during such cases ? 
if infact you can return right away if its empty.

Reply:
> active timeline reload is an expensive operation. we should try to minimize. can you check if exporedInstants is not empty and reload only during such cases ? if infact you can return right away if its empty.

Nice Suggestion. Already fixed it

Reply:
> Yes，we use a separate spark job to run table services and the instant in completed by the flink writer

But from the code, the compaction and clustering instants are excluded, can you elaborate a little more, what kind of table service the Spark offline job takes over?

Reply:
> > Yes，we use a separate spark job to run table services and the instant in completed by the flink writer
> 
> But from the code, the compaction and clustering instants are excluded, can you elaborate a little more, what kind of table service the Spark offline job takes over?

We use spark offline job  to run clean/archive, and lazy clean will rollback failed instants